<!DOCTYPE html>
<html lang="en"><head>
<script src="introduction_to_transcriptomics_files/libs/clipboard/clipboard.min.js"></script>
<script src="introduction_to_transcriptomics_files/libs/quarto-html/tabby.min.js"></script>
<script src="introduction_to_transcriptomics_files/libs/quarto-html/popper.min.js"></script>
<script src="introduction_to_transcriptomics_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="introduction_to_transcriptomics_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="introduction_to_transcriptomics_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="introduction_to_transcriptomics_files/libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.7.32">

  <meta name="author" content="Dr Stevie Pederson (They/Them)  stevie.pederson@thekids.org.au">
  <title>Introduction To Transcriptomics</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="introduction_to_transcriptomics_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="introduction_to_transcriptomics_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="introduction_to_transcriptomics_files/libs/revealjs/dist/theme/quarto-6a216ba96c5c25e91d68b44b6d1a2940.css">
  <link href="introduction_to_transcriptomics_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="introduction_to_transcriptomics_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="introduction_to_transcriptomics_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="introduction_to_transcriptomics_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Introduction To Transcriptomics</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Dr Stevie Pederson (They/Them) <br>stevie.pederson@thekids.org.au 
</div>
        <p class="quarto-title-affiliation">
            Black Ochre Data Labs, The Kids Research Institute Australia
          </p>
    </div>
</div>

</section>
<section id="acknowledgement-of-country" class="slide level2">
<h2><span class="text-red">Acknowledgement Of Country</span></h2>
<div class="text-red">
<p>I’d like to acknowledge the Kaurna people as the traditional owners and custodians of the land we know today as the Adelaide Plains, where I live &amp; work.</p>
<p>I also acknowledge the deep feelings of attachment and relationship of the Kaurna people to their place.</p>
<p>I pay my respects to the cultural authority of Aboriginal and Torres Strait Islander peoples from other areas of Australia, and pay my respects to Elders past, present and emerging, and acknowledge any Aboriginal Australians who may be with us today</p>
</div>
</section>
<section>
<section id="what-is-transcriptomics" class="title-slide slide level1 center">
<h1>What Is Transcriptomics</h1>

</section>
<section id="what-is-transcriptomics-1" class="slide level2">
<h2>What Is Transcriptomics?</h2>
<blockquote>
<p>Transcription is the process of making an RNA copy of a gene sequence</p>
</blockquote>
<div class="fragment">
<div class="columns">
<div class="column" style="width:42%;">
<ul>
<li><p>DNA can be described as being like a giant book of instructions</p></li>
<li><p>Some regions are defined as <em>genes</em></p>
<ul>
<li>Originally considered to be the basic unit of inheritance</li>
<li>Now used to describe a region of DNA transcribed into RNA</li>
</ul></li>
</ul>
</div><div class="column" style="width:58%;">
<div style="font-size: 90%; text-align: center;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="assets/transcription.jpg"></p>
<figcaption>Figure taken from <span class="citation" data-cites="Anderson2016-jc">Anderson and Bartee (<a href="#/section" role="doc-biblioref" onclick="">2016</a>)</span></figcaption>
</figure>
</div>
</div>
</div></div>
</div>
</section>
<section id="what-is-transcriptomics-2" class="slide level2 slide-only unlisted">
<h2>What Is Transcriptomics?</h2>
<blockquote>
<p>Transcriptomics is the study of transcribed RNA</p>
</blockquote>
<div class="fragment">
<div class="columns">
<div class="column" style="width:52%;">
<p>Transcribed RNA falls into 2 broad classes</p>
<ol type="1">
<li>Messenger RNA (mRNA)
<ul>
<li>Codes for protein sequences</li>
</ul></li>
<li>Non-coding RNA (ncRNA)
<ul>
<li>Multiple types of <em>functional</em> RNA</li>
<li>rRNA, tRNA <span class="math inline">\(\implies\)</span> protein translation</li>
<li>lncRNA, miRNA, snRNA, piRNA etc</li>
</ul></li>
</ol>
<p>Prokaryotes <span class="math inline">\(\implies\)</span> mRNA, rRNA &amp; tRNA</p>
</div><div class="column" style="width:48%;">
<div style="font-size: 60%; text-align: center;">
<p><img data-src="assets/DNA_to_protein_or_ncRNA.png"> <br>By <a href="//commons.wikimedia.org/wiki/User:Evolution_and_evolvability" title="User:Evolution and evolvability">Thomas Shafee</a> - <span class="int-own-work" lang="en">Own work</span>, <a href="https://creativecommons.org/licenses/by/4.0" title="Creative Commons Attribution 4.0">CC BY 4.0</a>, <a href="https://commons.wikimedia.org/w/index.php?curid=39441809">Wikimedia Link</a></p>
</div>
</div></div>
</div>
</section>
<section id="the-rna-population-of-a-eukaryotic-cell" class="slide level2">
<h2>The RNA Population Of a Eukaryotic Cell</h2>
<div class="columns">
<div class="column" style="width:60%;">
<div class="cell" data-layout-align="left">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure>
<p><img data-src="assets/eukaryotic_rna_types.png" style="width:92.0%"></p>
<figcaption>Image taken from <span class="citation" data-cites="ijms19051310">Chan and Tay (<a href="#/section" role="doc-biblioref" onclick="">2018</a>)</span></figcaption>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:40%;">
<p><br></p>
<ul>
<li>rRNA <span class="math inline">\(\approx\)</span> 80%<sup>1</sup></li>
<li>tRNA <span class="math inline">\(\approx\)</span> 15%</li>
<li>All other RNA <span class="math inline">\(\approx\)</span> 5% <br><br></li>
</ul>
<div style="font-size:60%">
<ol type="1">
<li>https://bionumbers.hms.harvard.edu/bionumber.aspx?s=n&amp;v=5&amp;id=100264</li>
</ol>
</div>
</div></div>
</section>
<section id="functional-rna" class="slide level2">
<h2>Functional RNA</h2>
<div class="columns">
<div class="column" style="width:60%;">
<div class="quarto-figure quarto-figure-left">
<figure>
<p><img data-src="https://imgs.xkcd.com/comics/rna.png"></p>
<figcaption>https://m.xkcd.com/3056/</figcaption>
</figure>
</div>
</div><div class="column" style="width:40%;">
<p>(An incomplete list)</p>
<ul>
<li>pre-mRNA + mRNA</li>
<li>lncRNA + lincRNA</li>
<li>miRNA, siRNA, shRNA, piRNA</li>
<li>rRNA + tRNA</li>
<li>snRNA + snoRNA</li>
<li>SRP RNA</li>
<li>eRNA</li>
<li>circRNA</li>
</ul>
</div></div>
</section>
<section id="eukaryotic-mrna-processing" class="slide level2">
<h2>Eukaryotic mRNA Processing</h2>
<ul>
<li>Nuclear mRNA have 5’ cap added
<ul>
<li>Protects single-stranded mRNA from degradation</li>
<li>Regulates nuclear export</li>
<li>Promotes translation into protein</li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li>mRNAs are polyadenylated at the 3’ end (-AAAAAAAAAAAAA)
<ul>
<li>Also protects from degradation</li>
<li>Aids in transcription termination, export and translation</li>
</ul></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Introns are spliced out as required<br>
</li>
</ul>
</div>
</section>
<section id="eukaryotic-mrna-processing-1" class="slide level2 slide-only unlisted">
<h2>Eukaryotic mRNA Processing</h2>

<img data-src="assets/eukaryotic_mrna_processing.png" style="width:80.0%" class="r-stretch quarto-figure-left"><p class="caption">Taken from <span class="citation" data-cites="Shafee2017-zv">Shafee and Lowe (<a href="#/section" role="doc-biblioref" onclick="">2017</a>)</span></p></section>
<section id="eukaryotic-mrna-processing-2" class="slide level2 slide-only unlisted">
<h2>Eukaryotic mRNA Processing</h2>

<img data-src="assets/DNA_alternative_splicing.png" class="r-stretch quarto-figure-center"><p class="caption">Image by the National Human Genome Research Institute</p></section></section>
<section>
<section id="why-study-transcriptomics" class="title-slide slide level1 center">
<h1>Why Study Transcriptomics?</h1>

</section>
<section id="why-study-transcriptomics-1" class="slide level2">
<h2>Why Study Transcriptomics?</h2>
<ul>
<li>Is a snapshot of highly <strong>dynamic biological processes</strong>
<ul>
<li>Captures response to stimulus and steady-state dynamics</li>
</ul></li>
<li>Assumed to be low-level
<ul>
<li>DNA <span class="math inline">\(\rightarrow\)</span> <span class="text-red">RNA</span> <span class="math inline">\(\rightarrow\)</span> Protein <span class="math inline">\(\rightarrow\)</span> Metabolites, Signalling molecules, etc …</li>
</ul></li>
<li>Use to make inference about these biological processes of interest
<ul>
<li>Can infer specific cell-cell communication methods</li>
<li>Identify therapeutic targets for Cardiovascular Disease, biomarkers for CAR-T cells etc</li>
</ul></li>
</ul>
</section>
<section id="quantitative-approaches" class="slide level2">
<h2>Quantitative Approaches</h2>
<ul>
<li>RNA expression is a rapid, early response to stimuli
<ul>
<li>Could be immune signalling, drug treatment etc</li>
</ul></li>
<li>Also changes in steady-state over time
<ul>
<li>First trimester placenta is hypoxic <span class="math inline">\(\implies\)</span> later is normoxic</li>
</ul></li>
<li>Change in a gene’s transcriptional activity <span class="math inline">\(\implies\)</span> change in RNA abundance
<ul>
<li>Capturing changes in abundance <span class="math inline">\(\implies\)</span> measure RNA quantities</li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li>Expression patterns <span class="math inline">\(\implies\)</span> identify cell-types in a heterogeneous sample</li>
<li>Changes in splicing patterns
<ul>
<li>Require methods for quantifying isoforms <em>within</em> a gene</li>
<li>May be changing proportions within gene-level abundances</li>
</ul></li>
</ul>
</div>
</section>
<section id="sequence-based-approaches" class="slide level2">
<h2>Sequence Based Approaches</h2>
<ul>
<li>Identify novel transcript sequences</li>
<li>No reference genome/transcriptome
<ul>
<li>Compare novel sequences against known transcriptomes <span class="math inline">\(\implies\)</span> infer function</li>
</ul></li>
<li>Sequences may diverge from reference
<ul>
<li>Do SNPs + InDels impact splicing/expression patterns in individual organisms</li>
</ul></li>
<li>Unexpected splicing patterns and chimeric RNA
<ul>
<li>Real genomes are less “neat” than reference genomes</li>
<li>Can play a key role in leukaemias &amp; other cancers <span class="math inline">\(\implies\)</span> clinical diagnostic</li>
</ul></li>
</ul>
</section></section>
<section>
<section id="the-development-of-transcriptomics" class="title-slide slide level1 center">
<h1>The Development of Transcriptomics</h1>

</section>
<section id="early-transcriptomics" class="slide level2">
<h2>Early Transcriptomics</h2>
<ul>
<li>The field developed with few reference sequences
<ul>
<li>Human Genome Project (1990-2003)</li>
</ul></li>
<li>Single sequence methods
<ul>
<li>Quantitative: Northern Blot (1977) + qPCR (1996)</li>
<li>Sequence Identification: Sanger Sequencing (1977)</li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li>High-Throughput Era
<ul>
<li>Quantitation: SAGE (1995) <span class="math inline">\(\rightarrow\)</span> Microarrays (1996)</li>
<li>Sequence Identification: ESTs (1991)</li>
</ul></li>
</ul>
</div>
</section>
<section id="northern-blots" class="slide level2">
<h2>Northern Blots</h2>
<div class="columns">
<div class="column" style="width:65%;">
<div class="cell" data-layout-align="left">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure>
<p><img data-src="assets/Northern-blot.jpg" style="width:100.0%"></p>
<figcaption>Figure taken from https://www.genome.gov/genetics-glossary/Northern-Blot</figcaption>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:35%;">
<div style="font-size:85%;">
<ul>
<li>Probes require sequence knowledge</li>
<li>Clear Presence/Absence calls</li>
<li>Crude quantitation: Densitometric Analysis</li>
</ul>
</div>
</div></div>
</section>
<section id="rt-qpcr" class="slide level2">
<h2>RT-qPCR</h2>
<aside class="notes">
<p>The C<sub>T</sub> values is actually estimated to a decimal value</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<ul>
<li>“Gold-standard” for measurement of transcription levels
<ul>
<li>Single gene <span class="math inline">\(\implies\)</span> not a high-throughput technique</li>
</ul></li>
<li>Targets a single transcript region with specific primers to produce cDNA <br><span class="math inline">\(\rightarrow\)</span> Polymerase Chain Reaction (PCR)</li>
<li>Each PCR cycle approximately doubles the target region</li>
</ul>
<div class="fragment">
<ul>
<li>cDNA produced is identified using fluorophores
<ul>
<li>Fluorescence doubles with each cycle</li>
</ul></li>
<li>Once fluorescence passes a detection threshold, the cycle number is recorded
<ul>
<li>Known as the Cycle Threshold (C<sub>T</sub>) value</li>
</ul></li>
</ul>
</div>
</section>
<section id="rt-qpcr-1" class="slide level2 slide-only unlisted">
<h2>RT-qPCR</h2>

<img data-src="assets/qPCR.png" style="width:70.0%" class="r-stretch quarto-figure-left"><p class="caption">A 10-fold dilution series</p></section>
<section id="rt-qpcr-2" class="slide level2 slide-only unlisted">
<h2>RT-qPCR</h2>
<ul>
<li>Higher C<sub>T</sub> values <span class="math inline">\(\implies\)</span> lower numbers of target molecule at the beginning</li>
<li>These can be used to estimate and compare abundance levels (i.e.&nbsp;gene expression)</li>
</ul>
<div class="fragment">
<ul>
<li>Is vulnerable to technical artefacts (e.g.&nbsp;pipetting &amp; sample variability)</li>
<li>Often includes one or more “housekeeper” genes thought to be stably expressed</li>
</ul>
</div>
<ul>
<li class="fragment">C<sub>T</sub> values are <em>normalised</em> to the housekeeper genes <span class="math inline">\(\implies C_{T_{hk}}\)</span>
<ul>
<li class="fragment">log<sub>2</sub> transformed values are used: <span class="math inline">\(\Delta C_T = \log_2 C_{T_g} - \log_2 C_{T_{hk}}\)</span></li>
</ul></li>
<li class="fragment">Change between conditions is the <em>change in</em> <span class="math inline">\(\Delta C_T \implies \Delta\Delta C_T\)</span></li>
<li class="fragment">Represents change on the log<sub>2</sub> scale, i.e.&nbsp;<em>log fold-change</em></li>
</ul>
</section>
<section id="expressed-sequence-tags-ests" class="slide level2">
<h2>Expressed Sequence Tags (ESTs)</h2>
<aside class="notes">
<ul>
<li>The senior author on the EST paper was J Craig Ventner who played an important role in the Human Genome Project</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<ul>
<li>The first attempt at capturing the larger transcriptome was ESTs <span class="citation" data-cites="1991VenterEST">(<a href="#/section" role="doc-biblioref" onclick="">Adams et al. 1991</a>)</span></li>
<li>Identified 609 human brain mRNA sequences
<ul>
<li>Selected for polyA-mRNA then reverse transcribed</li>
<li>Used random primers <span class="math inline">\(\rightarrow\)</span> Sanger Sequencing</li>
</ul></li>
<li>10 years before the Human Genome Project
<ul>
<li>Gene discovery was a hot topic</li>
</ul></li>
</ul>
</section>
<section id="serial-analysis-of-gene-expression-sage" class="slide level2">
<h2>Serial Analysis of Gene Expression (SAGE)</h2>
<ul>
<li>First high-throughput quantification method was <em>Serial Analysis of Gene Expression</em> (SAGE) <span class="citation" data-cites="pmid7570003">(<a href="#/section" role="doc-biblioref" onclick="">Velculescu et al. 1995</a>)</span></li>
</ul>
<div class="columns">
<div class="column" style="width:42%;">
<div style="font-size: 80%; text-align: center;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="assets/SAGE.png"></p>
<figcaption>Thomas Shafee, <a href="https://creativecommons.org/licenses/by/4.0">CC BY 4.0</a>, via Wikimedia Commons</figcaption>
</figure>
</div>
</div>
</div><div class="column" style="width:58%;">
<ul>
<li>mRNA <span class="math inline">\(\rightarrow\)</span> cDNA using biotinylated primers</li>
<li>cDNA bound to beads (using biotin) &amp; cleaved</li>
<li>11mer “tags” were ligated into long sequences using linker sequences</li>
<li>Sequenced using Sanger Sequencing</li>
<li>Deconvolution &amp; counting</li>
</ul>
<ul>
<li class="fragment">First count-based transcriptomic methods developed</li>
</ul>
</div></div>
</section>
<section id="microarray-technology" class="slide level2">
<h2>Microarray Technology</h2>
<div style="font-size:90%">
<div class="columns">
<div class="column" style="width:40%;">
<div class="quarto-figure quarto-figure-left">
<figure>
<p><img data-src="assets/2colour.jpg" style="width:100.0%"></p>
<figcaption>Section of two-colour array taken from <span class="citation" data-cites="Shalon01071996">Shalon, Smith, and Brown (<a href="#/section" role="doc-biblioref" onclick="">1996</a>)</span></figcaption>
</figure>
</div>
</div><div class="column" style="width:60%;">
<ul>
<li>Truly launched the modern transcriptomics era</li>
<li>Quantified thousands of transcripts simultaneously</li>
<li>Relied on development of Human Genome Project (+ other organisms)</li>
<li>Analysis in R/Bioconductor
<ul>
<li>Rv1.0.0 (2000)</li>
<li>Bioconductor <span class="citation" data-cites="Gentleman2004-sd">(<a href="#/section" role="doc-biblioref" onclick="">Gentleman et al. 2004</a>)</span></li>
<li>Modern statistical high-throughput models developed</li>
</ul></li>
</ul>
</div></div>
</div>
</section></section>
<section>
<section id="references" class="title-slide slide level1 center">
<h1>References</h1>

</section>
<section id="section" class="slide level2 smaller scrollable">
<h2></h2>
<!-- (no special size adjustment for HTML) -->
<!-- This is where the bibliography gets injected -->
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-1991VenterEST" class="csl-entry" role="listitem">
Adams, Mark D., Jenny M. Kelley, Jeannine D. Gocayne, Mark Dubnick, Mihael H. Polymeropoulos, Hong Xiao, Carl R. Merril, et al. 1991. <span>“Complementary DNA Sequencing: Expressed Sequence Tags and Human Genome Project.”</span> <em>Science</em> 252 (5013): 1651–56. <a href="http://www.jstor.org/stable/2876333">http://www.jstor.org/stable/2876333</a>.
</div>
<div id="ref-Anderson2016-jc" class="csl-entry" role="listitem">
Anderson, Christine, and Lisa Bartee. 2016. <em>Mt Hood Community College Biology 102</em>. Open Oregon Educational Resources.
</div>
<div id="ref-ijms19051310" class="csl-entry" role="listitem">
Chan, Jia Jia, and Yvonne Tay. 2018. <span>“Noncoding RNA:RNA Regulatory Networks in Cancer.”</span> <em>International Journal of Molecular Sciences</em> 19 (5). <a href="https://doi.org/10.3390/ijms19051310">https://doi.org/10.3390/ijms19051310</a>.
</div>
<div id="ref-Gentleman2004-sd" class="csl-entry" role="listitem">
Gentleman, Robert C, Vincent J Carey, Douglas M Bates, Ben Bolstad, Marcel Dettling, Sandrine Dudoit, Byron Ellis, et al. 2004. <span>“Bioconductor: Open Software Development for Computational Biology and Bioinformatics.”</span> <em>Genome Biol.</em> 5 (10): R80.
</div>
<div id="ref-Shafee2017-zv" class="csl-entry" role="listitem">
Shafee, Thomas, and Rohan Lowe. 2017. <span>“Eukaryotic and Prokaryotic Gene Structure.”</span> <em><span>WikiJournal</span> of Medicine</em>, January. <a href="https://doi.org/10.15347/WJM/2017.002">https://doi.org/10.15347/WJM/2017.002</a>.
</div>
<div id="ref-Shalon01071996" class="csl-entry" role="listitem">
Shalon, D, S J Smith, and P O Brown. 1996. <span>“A DNA Microarray System for Analyzing Complex DNA Samples Using Two-Color Fluorescent Probe Hybridization.”</span> <em>Genome Research</em> 6 (7): 639–45. <a href="https://doi.org/10.1101/gr.6.7.639">https://doi.org/10.1101/gr.6.7.639</a>.
</div>
<div id="ref-pmid7570003" class="csl-entry" role="listitem">
Velculescu, V. E., L. Zhang, B. Vogelstein, and K. W. Kinzler. 1995. <span>“<span class="nocase"><span>S</span>erial analysis of gene expression</span>.”</span> <em>Science</em> 270 (5235): 484–87.
</div>
</div>

</section></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="introduction_to_transcriptomics_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="introduction_to_transcriptomics_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="introduction_to_transcriptomics_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="introduction_to_transcriptomics_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="introduction_to_transcriptomics_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="introduction_to_transcriptomics_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="introduction_to_transcriptomics_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="introduction_to_transcriptomics_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="introduction_to_transcriptomics_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="introduction_to_transcriptomics_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    <div id="logo-tki" style="position: fixed; z-index: 1000;">
      <img src="https://www.thekids.org.au/Static/Theme/images/TheKids-RGB_Logo_Colour-Pattern.svg" style="height: 80px;">
    </div>
    <div id="logo-bodl" style="position: fixed; z-index: 1000;">
      <img src="assets/bodl_logo_white_background.jpg" style="height: 80px;">
    </div>
    <div id="logo-uoa" style="z-index: 1000;">
      <img src="assets/UoA_logo_col_vert.png" style="height: 80px;">
    </div>


    <script>
    function positionLogos() {
      const tkiLogo = document.getElementById('logo-tki');
      const uoaLogo = document.getElementById('logo-uoa');
      const bodlLogo = document.getElementById('logo-bodl');

      // Check if this is RevealJS
      const isRevealJS = document.body.classList.contains('reveal') ||
                         document.querySelector('.reveal') !== null ||
                         typeof Reveal !== 'undefined';

      if (isRevealJS) {
        // RevealJS - both logos at bottom
        tkiLogo.style.bottom = '10px';
        tkiLogo.style.right = '340px';
        tkiLogo.style.top = 'auto';
        tkiLogo.style.display = 'block';

        bodlLogo.style.bottom = '10px';
        bodlLogo.style.right = '160px';
        bodlLogo.style.top = 'auto';
        bodlLogo.style.display = 'block';

        uoaLogo.style.bottom = '10px';
        uoaLogo.style.right = '40px';
        uoaLogo.style.top = 'auto';
        uoaLogo.style.display = 'block';
        uoaLogo.style.position = 'fixed';
      } else {
        // Flat HTML - only right logo at top
        tkiLogo.style.display = 'none';
        bodlLogo.style.display = 'none';

        uoaLogo.style.top = '30px';
        uoaLogo.style.left = '60%';
        uoaLogo.style.bottom = 'auto';
        uoaLogo.style.position = 'absolute';
        uoaLogo.style.display = 'block';
      }
    }

    // Run positioning
    document.addEventListener('DOMContentLoaded', positionLogos);
    setTimeout(positionLogos, 100);
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
            const codeEl = trigger.previousElementSibling.cloneNode(true);
            for (const childEl of codeEl.children) {
              if (isCodeAnnotation(childEl)) {
                childEl.remove();
              }
            }
            return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp('/' + window.location.host + '/');
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    

</body></html>