---
title: "Introduction To Transcriptomics"
author: "Dr Stevie Pederson (They/Them) <br>stevie.pederson@thekids.org.au"
email: "stevie.pederson@thekids.org.au"
institute: "Black Ochre Data Labs, The Kids Research Institute Australia"
bibliography: assets/references.bib
format: 
  revealjs:
    theme: [assets/custom.scss]
    output-file: introduction_to_transcriptomics-slides.html
    height: 720
    width: 1280
    sansfont: Times New Roman
    slide-number: c
    show-slide-number: all
    include-after-body: assets/uoa_bodl_logos.html
  html: 
    css: [assets/custom.scss]
    output-file: introduction_to_transcriptomics.html
    embed-resources: true    
    toc: true
    toc-depth: 2
    include-in-header: assets/uoa_bodl_logos.html
  beamer-pdf:
    output-file: introduction_to_transcriptomics.pdf
    theme: Pittsburgh
    include-in-header: assets/beamer-logos.tex
    geometry:
      - paperwidth=16cm
      - paperheight=9cm
      - margin=0.8cm
knitr: 
  opts_chunk: 
    set:
      echo: false
      fig.align: 'left'
---

## [Acknowledgement Of Country]{.text-red}

::: {.text-red}

I’d like to acknowledge the Kaurna people as the traditional owners and custodians of the land we know today as the Adelaide Plains, where I live & work.

I also acknowledge the deep feelings of attachment and relationship of the Kaurna people to their place.

I pay my respects to the cultural authority of Aboriginal and Torres Strait Islander peoples from other areas of Australia, and pay my respects to Elders past, present and emerging, and acknowledge any Aboriginal Australians who may be with us today

:::

# What Is Transcriptomics

## What Is Transcriptomics?

> Transcription is the process of making an RNA copy of a gene sequence

::: {.fragment}

:::: {.columns}

::: {.column width='42%'}


- DNA can be described as being like a giant book of instructions

- Some regions are defined as *genes*
    + Originally considered to be the basic unit of inheritance
    + Now used to describe a region of DNA transcribed into RNA
    
:::

::: {.column width='58%'}

::: {style="font-size: 90%; text-align: center;"}

![Figure taken from @Anderson2016-jc](assets/transcription.jpg)

:::

:::

::::

:::

[^1]: https://openoregon.pressbooks.pub/mhccbiology102/chapter/transcription/

## What Is Transcriptomics? {.slide-only .unlisted}

> Transcriptomics is the study of transcribed RNA

::: {.fragment}

:::: {.columns}

::: {.column width='52%'}

Transcribed RNA falls into 2 broad classes

1. Messenger RNA (mRNA) 
    + Codes for protein sequences
2. Non-coding RNA (ncRNA)
    + Multiple types of *functional* RNA
    + rRNA, tRNA $\implies$ protein translation
    + lncRNA, miRNA, snRNA, piRNA etc
    
Prokaryotes $\implies$ mRNA, rRNA & tRNA

::: {.content-visible when-format="beamer"}
\vspace{1em}
:::



:::
::: {.column width='48%'}


::: {style="font-size: 60%; text-align: center;"}

![](assets/DNA_to_protein_or_ncRNA.png)
<br>By <a href="//commons.wikimedia.org/wiki/User:Evolution_and_evolvability" title="User:Evolution and evolvability">Thomas Shafee</a> - <span class="int-own-work" lang="en">Own work</span>, <a href="https://creativecommons.org/licenses/by/4.0" title="Creative Commons Attribution 4.0">CC BY 4.0</a>, <a href="https://commons.wikimedia.org/w/index.php?curid=39441809">Wikimedia Link</a>

:::
:::

::::

:::

## The RNA Population Of a Eukaryotic Cell

:::: {.columns}

::: {.column width='60%'}

```{r rna-types, echo = FALSE, fig.cap = "Image taken from @ijms19051310", out.width='92%', fig.align='left'}
knitr::include_graphics("assets/eukaryotic_rna_types.png")
```



:::
::: {.column width='40%'}

<br>

::: {.content-visible when-format="beamer"}
\vspace{2em}
:::

- rRNA $\approx$ 80%^1^
- tRNA $\approx$ 15% 
- All other RNA $\approx$ 5%
<br><br>

::: {.content-visible when-format="revealjs"}
::: {style="font-size:60%"}
1. https://bionumbers.hms.harvard.edu/bionumber.aspx?s=n&v=5&id=100264

:::
:::

::: {.content-visible when-format="beamer"}
\vspace{2em}
\tiny ^1^ https://bionumbers.hms.harvard.edu/bionumber.aspx
:::


:::

::::

## Functional RNA

:::: {.columns}

::: {.column width='60%'}

![https://m.xkcd.com/3056/](https://imgs.xkcd.com/comics/rna.png){fig-align="left"}
:::

::: {.column width='40%'}
(An incomplete list)

- pre-mRNA + mRNA
- lncRNA + lincRNA
- miRNA, siRNA, shRNA, piRNA
- rRNA + tRNA
- snRNA + snoRNA
- SRP RNA
- eRNA
- circRNA
:::

::::



## Eukaryotic mRNA Processing

- Nuclear mRNA have 5' cap added
    + Protects single-stranded mRNA from degradation
    + Regulates nuclear export
    + Promotes translation into protein
    
::: {.fragment}
- mRNAs are polyadenylated at the 3' end (-AAAAAAAAAAAAA)
    + Also protects from degradation
    + Aids in transcription termination, export and translation
    
:::
::: {.fragment}
- Introns are spliced out as required    
:::


## Eukaryotic mRNA Processing {.slide-only .unlisted}

```{r mrna-processing, echo = FALSE, fig.cap="Taken from @Shafee2017-zv", fig.align='left', out.width='80%'}
knitr::include_graphics(here::here("assets/eukaryotic_mrna_processing.png"))
```

## Eukaryotic mRNA Processing {.slide-only .unlisted}

![Image by the National Human Genome Research Institute](assets/DNA_alternative_splicing.png)

# Why Study Transcriptomics? 

## Why Study Transcriptomics? 

- Is a snapshot of highly **dynamic biological processes**
    + Captures response to stimulus and steady-state dynamics
- Assumed to be low-level
    + DNA $\rightarrow$ [RNA]{.text-red} $\rightarrow$ Protein $\rightarrow$ Metabolites, Signalling molecules, etc ...
- Use to make inference about these biological processes of interest
    + Can infer specific cell-cell communication methods
    + Identify therapeutic targets for Cardiovascular Disease, biomarkers for CAR-T cells etc

## Quantitative Approaches

- RNA expression is a rapid, early response to stimuli
    + Could be immune signalling, drug treatment etc
- Also changes in steady-state over time
    + First trimester placenta is hypoxic $\implies$ later is normoxic
- Change in a gene's transcriptional activity $\implies$ change in RNA abundance
    + Capturing changes in abundance $\implies$ measure RNA quantities

::: {.fragment}
    
- Expression patterns $\implies$ identify cell-types in a heterogeneous sample
- Changes in splicing patterns
    + Require methods for quantifying isoforms *within* a gene
    + May be changing proportions within gene-level abundances

:::
    
## Sequence Based Approaches

- Identify novel transcript sequences
- No reference genome/transcriptome
    + Compare novel sequences against known transcriptomes $\implies$ infer function
- Sequences may diverge from reference
    + Do SNPs + InDels impact splicing/expression patterns in individual organisms
- Unexpected splicing patterns and chimeric RNA
    + Real genomes are less "neat" than reference genomes
    + Can play a key role in leukaemias & other cancers $\implies$ clinical diagnostic

# The Development of Transcriptomics

## Early Transcriptomics

- The field developed with few reference sequences
    + Human Genome Project (1990-2003)
- Single sequence methods
    + Quantitative: Northern Blot (1977) + qPCR (1996)
    + Sequence Identification: Sanger Sequencing (1977)
    
::: {.fragment}
    
- High-Throughput Era
    + Quantitation: SAGE (1995) $\rightarrow$ Microarrays (1996)
    + Sequence Identification: ESTs (1991)

:::

## Northern Blots

:::: {.columns}
::: {.column width='65%'}

```{r northern-blot, echo = FALSE, out.width = '100%', fig.align = 'left', fig.cap = "Figure taken from https://www.genome.gov/genetics-glossary/Northern-Blot"}
knitr::include_graphics(here::here("assets/Northern-blot.jpg"))
```

:::

::: {.column width='35%'}
::: {style='font-size:85%;'}
- Probes require sequence knowledge
- Clear Presence/Absence calls
- Crude quantitation: Densitometric Analysis
:::
:::

::::

## RT-qPCR

::: {.notes}
The C~T~ values is actually estimated to a decimal value
:::

- “Gold-standard” for measurement of transcription levels
    + Single gene $\implies$ not a high-throughput technique
- Targets a single transcript region with specific primers to produce cDNA <br>$\rightarrow$ Polymerase Chain Reaction (PCR)
- Each PCR cycle approximately doubles the target region
    
::: {.fragment}    
- cDNA produced is identified using fluorophores
    + Fluorescence doubles with each cycle
- Once fluorescence passes a detection threshold, the cycle number is recorded
    + Known as the Cycle Threshold (C~T~) value

:::

## RT-qPCR {.slide-only .unlisted}

```{r rt-pcr, echo = FALSE, out.width = '70%', fig.align = 'left', fig.cap = "A 10-fold dilution series"}
knitr::include_graphics(here::here("assets/qPCR.png"))
```



## RT-qPCR {.slide-only .unlisted}

- Higher C~T~ values $\implies$ lower numbers of target molecule at the beginning
- These can be used to estimate and compare abundance levels (i.e. gene expression)

::: {.fragment}
- Is vulnerable to technical artefacts (e.g. pipetting & sample variability)
- Often includes one or more "housekeeper" genes thought to be stably expressed
:::

::: {.incremental}
- C~T~ values are *normalised* to the housekeeper genes $\implies C_{T_{hk}}$
    + log~2~ transformed values are used: $\Delta C_T = \log_2 C_{T_g} - \log_2 C_{T_{hk}}$
- Change between conditions is the *change in* $\Delta C_T \implies \Delta\Delta C_T$
- Represents change on the log~2~ scale, i.e. *log fold-change*
:::

## Expressed Sequence Tags (ESTs)

::: {.notes}
- The senior author on the EST paper was J Craig Ventner who played an important role in the Human Genome Project
:::

- The first attempt at capturing the larger transcriptome was ESTs [@1991VenterEST]
- Identified 609 human brain mRNA sequences
    + Selected for polyA-mRNA then reverse transcribed
    + Used random primers $\rightarrow$ Sanger Sequencing
- 10 years before the Human Genome Project
    + Gene discovery was a hot topic

## Serial Analysis of Gene Expression (SAGE)

- First high-throughput quantification method was *Serial Analysis of Gene Expression* (SAGE) [@pmid7570003]

:::: {.columns}

::: {.column width="42%"}
::: {style="font-size: 80%; text-align: center;"}
![Thomas Shafee, [CC BY 4.0](https://creativecommons.org/licenses/by/4.0), via Wikimedia Commons](assets/SAGE.png)
:::
:::

::: {.column width="58%"}

- mRNA $\rightarrow$ cDNA using biotinylated primers
- cDNA bound to beads (using biotin) & cleaved 
- 11mer "tags" were ligated into long sequences using linker sequences
- Sequenced using Sanger Sequencing
- Deconvolution & counting

::: {.incremental}
- First count-based transcriptomic methods developed
:::

:::

::::

## Microarray Technology

::: {style='font-size:90%'}

:::: {.columns}

::: {.column width='40%'}


![Section of two-colour array taken from @Shalon01071996](assets/2colour.jpg){width="100%" fig-align="left"}

:::

::: {.column width='60%'}

- Truly launched the modern transcriptomics era
- Quantified thousands of transcripts simultaneously
- Relied on development of Human Genome Project (+ other organisms)
- Analysis in R/Bioconductor
    + Rv1.0.0 (2000) 
    + Bioconductor [@Gentleman2004-sd]
    + Modern statistical high-throughput models developed
    
:::



::::

:::

# References

##

::: {.content-visible when-format="beamer"}
\begingroup
\scriptsize
:::

::: {.content-visible when-format="revealjs"}
<!-- (no special size adjustment for HTML) -->
:::

<!-- This is where the bibliography gets injected -->
:::{#refs}
:::

::: {.content-visible when-format="beamer"}
\endgroup
:::
