<!DOCTYPE html>
<html lang="en"><head>
<script src="rnaseq_files/libs/clipboard/clipboard.min.js"></script>
<script src="rnaseq_files/libs/quarto-html/tabby.min.js"></script>
<script src="rnaseq_files/libs/quarto-html/popper.min.js"></script>
<script src="rnaseq_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="rnaseq_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="rnaseq_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="rnaseq_files/libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.7.32">

  <meta name="author" content="Dr Stevie Pederson (They/Them)  stevie.pederson@thekids.org.au">
  <title>RNA Sequencing</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="rnaseq_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="rnaseq_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="rnaseq_files/libs/revealjs/dist/theme/quarto-6a216ba96c5c25e91d68b44b6d1a2940.css">
  <link href="rnaseq_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="rnaseq_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="rnaseq_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="rnaseq_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">RNA Sequencing</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Dr Stevie Pederson (They/Them) <br>stevie.pederson@thekids.org.au 
</div>
        <p class="quarto-title-affiliation">
            </p><div class="line-block">Black Ochre Data Labs<br>
            The Kids Research Institute Australia</div>
          <p></p>
    </div>
</div>

</section>
<section id="acknowledgement-of-country" class="slide level2">
<h2><span class="text-red">Acknowledgement Of Country</span></h2>
<div class="text-red">
<p>I’d like to acknowledge the Kaurna people as the traditional owners and custodians of the land we know today as the Adelaide Plains, where I live &amp; work.</p>
<p>I also acknowledge the deep feelings of attachment and relationship of the Kaurna people to their place.</p>
<p>I pay my respects to the cultural authority of Aboriginal and Torres Strait Islander peoples from other areas of Australia, and pay my respects to Elders past, present and emerging, and acknowledge any Aboriginal Australians who may be with us today</p>
</div>
</section>
<section id="rna-sequencing" class="slide level2">
<h2>RNA Sequencing</h2>
<p>According to <span class="citation" data-cites="Wang2009-hf">Wang, Gerstein, and Snyder (<a href="#/section" role="doc-biblioref" onclick="">2009</a>)</span></p>
<blockquote>
<p>RNA-Seq, also called RNA sequencing, is a particular technology-based sequencing technique which uses next-generation sequencing (NGS) to reveal the <strong>presence</strong> and <strong>quantity</strong> of RNA in a biological sample at a <strong>given moment</strong>, analyzing the continuously changing cellular transcriptome.</p>
</blockquote>
</section>
<section id="rna-sequencing-1" class="slide level2 slide-only unlisted">
<h2>RNA Sequencing</h2>
<div class="columns">
<div class="column">
<ul>
<li>Microarrays are still published regularly
<ul>
<li>Also used extensively for methylation</li>
</ul></li>
<li>RNA sequencing is now the dominant technology</li>
</ul>
<div class="fragment">
<ul>
<li>Strong improvement for:
<ul>
<li>transcript-level resolution</li>
<li>un-annotated genes</li>
<li>genomic variants</li>
<li>allelic bias</li>
</ul></li>
</ul>
</div>
</div><div class="column">
<div class="cell">
<div class="cell-output-display">
<div>
<figure>
<p><img data-src="rnaseq_files/figure-revealjs/microarray-vs-rnaseq-1.png" width="576"></p>
</figure>
</div>
</div>
</div>
</div></div>
</section>
<section id="rna-sequencing-2" class="slide level2 slide-only unlisted">
<h2>RNA Sequencing</h2>
<ul>
<li>Microarrays rely on probes for transcripts defined <em>at design time</em></li>
<li>Restricted in the number of transcripts/genes targeted
<ul>
<li>Arrays are confined for space</li>
<li>Gencode 48 (GRCh38): 78,686 genes + 385,669 transcripts</li>
</ul></li>
<li>Probes capture non-specific binding</li>
<li>Measuring a labelled cDNA: fluourescence <span class="math inline">\(\propto\)</span> RNA abundace</li>
</ul>
<div class="fragment">
<p><span class="text-red">These limitations do not exist for RNA-Seq</span></p>
</div>
</section>
<section id="rna-sequencing-3" class="slide level2 slide-only unlisted">
<h2>RNA Sequencing</h2>
<ul>
<li>Directly sequence the biological material</li>
<li>Map to most recent reference at any point in time</li>
<li>Assemble a transcriptome (tissue specific)</li>
<li>Detect InDels / SNPs in expressed sequences</li>
</ul>
<div class="fragment">
<ul>
<li>Multiple variations
<ul>
<li>total-RNA or polyA transcripts <span class="fragment"><span class="math inline">\(\implies\)</span> most similar to microarrays</span></li>
<li>small-RNA libraries</li>
<li>Long Reads (Oxford Nanopore, PacBio) <span class="fragment"><br><span class="math inline">\(\implies\)</span> originally isoform discovery, quantitative methods improving</span></li>
</ul></li>
</ul>
</div>
</section>
<section id="the-key-steps" class="slide level2">
<h2>The Key Steps</h2>
<ul>
<li>Focus from here on will be sequencing mRNA using short reads</li>
</ul>
<div class="columns">
<div class="column">
<div class="fragment">
<ol type="1">
<li>Library Preparation
<ul>
<li>RNA Quality assessment <br>(i.e.&nbsp;RNA degradation)</li>
<li>Selecting target molecules</li>
<li>Adding sequencing primers</li>
</ul></li>
</ol>
</div>
<div class="fragment">
<ol start="2" type="1">
<li>Sequencing</li>
<li>Alignment</li>
</ol>
</div>
</div><div class="column">
<div class="fragment">
<ol start="4" type="1">
<li>Quantitation (i.e.&nbsp;counting)</li>
<li>DE Gene Detection</li>
<li>Downstream Analysis</li>
</ol>
</div>
</div></div>
</section>
<section id="polya-based-rna-selection" class="slide level2">
<h2>PolyA-Based RNA Selection</h2>
<ol type="1">
<li>Select for poly-adenylated RNA using oligo-dT-based methods
<ul>
<li>Only extracts intact mRNA with a polyA tail (includes some ncRNA)</li>
</ul></li>
</ol>

<img data-src="assets/03polya_workflow1.jpg" style="width:80.0%" class="r-stretch quarto-figure-center"><p class="caption">Image from https://www.lexogen.com/polya-rna-selection-kit/</p></section>
<section id="rna-selection-via-rrna-depletion" class="slide level2">
<h2>RNA Selection via rRNA Depletion</h2>
<ol start="2" type="1">
<li>Enzymatically deplete rRNA sequences
<ul>
<li>rRNA targeted using probes <span class="math inline">\(\implies\)</span> dsRNA degraded</li>
<li>Can additionally target hbRNA (whole blood)</li>
</ul></li>
</ol>

<img data-src="assets/Ribo_Zero_1.jpg" style="width:70.0%" class="r-stretch quarto-figure-center"><p class="caption">Image from https://support.illumina.com.cn</p></section>
<section id="library-preparation" class="slide level2">
<h2>Library Preparation</h2>
<ul>
<li>RNA is then fragmented and size selected (200-300nt)
<ul>
<li>Very short transcripts always lost during this step</li>
</ul></li>
<li>cDNA produced</li>
<li>Sequencing adapters added
<ul>
<li>Indexes are unique to each individual library <span class="math inline">\(\implies\)</span> always have replicates</li>
<li>Optionally contain Unique Molecular Identifiers (UMI)<br><span class="math inline">\(\implies\)</span> Helps identify PCR duplicates</li>
</ul></li>
<li>Most RNA-Seq now retains <em>strand-of-origin</em> information (Stranded RNA-Seq)
<ul>
<li>During PCR only the first cDNA template retained</li>
</ul></li>
</ul>
</section>
<section id="library-preparation-1" class="slide level2 slide-only unlisted">
<h2>Library Preparation</h2>

<img data-src="assets/NGS-adapters_01.png" class="r-stretch quarto-figure-center"><p class="caption">Image courtesy of https://www.lubio.ch/blog/ngs-adapters</p></section>
<section id="library-preparation-2" class="slide level2 slide-only unlisted">
<h2>Library Preparation</h2>
<ul>
<li>Showing the full Y-adapter: Different combinations of indexes, UMI optional</li>
</ul>

<img data-src="assets/xgen-udi-umi-adapter-workflow.png" style="width:95.0%" class="r-stretch quarto-figure-center"><p class="caption">Image courtesy of https://sg.idtdna.com</p></section>
<section id="sequencing" class="slide level2">
<h2>Sequencing</h2>
<div class="columns">
<div class="column" style="width:65%;">
<div class="quarto-figure quarto-figure-left">
<figure>
<p><img data-src="assets/A8.jpg"></p>
<figcaption>Image taken from https://www.bmkgene.com/dnarna-sequencing-illumina-sequencer-product/</figcaption>
</figure>
</div>
</div><div class="column" style="width:35%;">
<ul>
<li>Standard Illumina short-read sequencing protocol shown</li>
<li>MGI uses different technology</li>
</ul>
</div></div>
</section>
<section>
<section id="alignment-to-a-reference-genome" class="title-slide slide level1 center">
<h1>Alignment To A Reference Genome</h1>

</section>
<section id="reference-genome-alignment" class="slide level2">
<h2>Reference Genome Alignment</h2>
<ul>
<li>Sequencing data arrives as FastQ files
<ul>
<li>Standard QC (<code>FastQC</code>/<code>fastp</code>)</li>
<li>Optional Adapter Removal</li>
</ul></li>
<li>Alignment to a reference genome needs to be splice aware
<ul>
<li>Usually indexed using a set of gene annotations</li>
<li>Indexes are dependent on both genome version + annotation version</li>
</ul></li>
<li>Most common aligners are STAR <span class="citation" data-cites="Dobin2013-us">(<a href="#/section" role="doc-biblioref" onclick="">Dobin et al. 2013</a>)</span> &amp; hisat2 <span class="citation" data-cites="Kim2019-ib">(<a href="#/section" role="doc-biblioref" onclick="">Kim et al. 2019</a>)</span>
<ul>
<li>Alignments returned as SAM/BAM files</li>
</ul></li>
</ul>
</section>
<section id="gene-annotations" class="slide level2">
<h2>Gene Annotations</h2>
<ul>
<li>Gene models are now well-annotated for model organisms
<ul>
<li>US maintained: RefSeq; UCSC (https://genome.ucsc.edu)</li>
<li>European maintained: EnsEMBL (https://www.ensembl.org/)</li>
</ul></li>
<li>Contain a mix of predicted and observed gene-models</li>
<li>Human/Mouse also use Gencode (https://www.gencodegenes.org)</li>
<li>Usually provided as GTF/GFF file
<ul>
<li>Contains exon-, transcript- and gene-level annotations</li>
<li>Co-ordinate-based, not sequence-based</li>
</ul></li>
</ul>
</section>
<section id="gene-quantitation" class="slide level2">
<h2>Gene Quantitation</h2>
<ul>
<li>After alignment to a reference gene <span class="math inline">\(\rightarrow\)</span> count reads
<ul>
<li>Leading to Differential Gene Expression analysis</li>
</ul></li>
<li>Multiple high-quality tools: <code>RSEM</code> <span class="citation" data-cites="Li2011-iu">(<a href="#/section" role="doc-biblioref" onclick="">Li and Dewey 2011</a>)</span>, <code>featureCounts</code> <span class="citation" data-cites="Liao2014-gy">(<a href="#/section" role="doc-biblioref" onclick="">Liao, Smyth, and Shi 2014</a>)</span>, <code>htseq</code> <span class="citation" data-cites="htseq2014">(<a href="#/section" role="doc-biblioref" onclick="">Anders, Pyl, and Huber 2014</a>)</span>
<ul>
<li>Always use the same GTF used when indexing the genome</li>
</ul></li>
<li>Easy in theory but biology is often inconvenient
<ul>
<li>Most alignments will neatly be within exons or across splice junctions</li>
<li>Some are not</li>
</ul></li>
</ul>
</section>
<section id="gene-quantitation-1" class="slide level2 slide-only unlisted">
<h2>Gene Quantitation</h2>
<div class="columns">
<div class="column">
<div class="cell">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="assets/count_modes.png" style="width:90.0%"></p>
<figcaption>Image taken from https://htseq.readthedocs.io/en/latest/htseqcount.html</figcaption>
</figure>
</div>
</div>
</div>
</div><div class="column">
<p>How do we count:</p>
<ul>
<li>reads that only partially overlap an exon</li>
<li>reads that map to multiple-locations (i.e.&nbsp;multi-mapping)</li>
<li>reads where genes are on sense-antisense strand</li>
<li>non-canonical splicing events</li>
</ul>
</div></div>
</section>
<section id="gene-quantitation-2" class="slide level2 slide-only unlisted">
<h2>Gene Quantitation</h2>
<ul>
<li>The region encoding a gene is (relatively) well defined
<ul>
<li>An alignment within a gene is easy to assign to that gene</li>
<li>Much more difficult to identify which transcript it came from</li>
</ul></li>
</ul>
<div class="fragment">
<div class="columns">
<div class="column" style="width:65%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="assets/gkv608fig1.jpeg"></p>
<figcaption>Image taken from <span class="citation" data-cites="Yu2015-f5">Yu et al. (<a href="#/section" role="doc-biblioref" onclick="">2015</a>)</span></figcaption>
</figure>
</div>
</div><div class="column" style="width:35%;">
<ul>
<li>Many transcripts share multiple exons</li>
<li>Splice Junctions were the earliest approach</li>
</ul>
</div></div>
</div>
</section>
<section id="count-based-data" class="slide level2">
<h2>Count-Based Data</h2>
<ul>
<li>For RNA-Seq: number of reads aligned to a gene <span class="math inline">\(\rightarrow\)</span> <em>gene expression</em>
<ul>
<li>Longer genes will return higher counts <span class="math inline">\(\implies\)</span> not observed in microarrays</li>
</ul></li>
<li>These are <em>discrete</em> data (i.e.&nbsp;not continuous values)</li>
<li>Microarrays were continuous values (fluorescence intensity)
<ul>
<li>Modelled using log<sub>2</sub>-transformed values <span class="math inline">\(\implies \mathcal{N}(\mu, \sigma)\)</span></li>
<li>Linear regression, <span class="math inline">\(t\)</span>-tests etc</li>
<li>Mean and variance are independent variables</li>
</ul></li>
</ul>
</section>
<section id="count-based-data-1" class="slide level2 slide-only unlisted">
<h2>Count-Based Data</h2>
<ul>
<li>Count data is commonly modelled using a Poisson Distribution <span class="math inline">\(\implies \text{Poisson}(\lambda)\)</span>
<ul>
<li>Good example is number of phone calls per minute at a signal tower</li>
<li>Cars per hour in an intersection</li>
</ul></li>
<li>Poisson Distributions define the variance as being equal to the mean
<ul>
<li>i.e.&nbsp;<span class="math inline">\(\sigma^2 = \mu = \lambda \implies\)</span>Mean and variance <em>are not independent variables</em></li>
<li><del>Linear regression, <span class="math inline">\(t\)</span>-tests etc</del> Generalised Linear Models (GLMs)</li>
</ul></li>
</ul>
<div class="fragment" data-fragment-index="1">
<ul>
<li>Biology is inconvenient <span class="math inline">\(\implies\)</span> for RNA-Seq counts <span class="math inline">\(\sigma^2 &gt; \mu\)</span> <span class="fragment" data-fragment-index="2"><span class="math inline">\(\implies\)</span> <em>overdispersion</em></span>
<ul>
<li>Need a different model <span class="fragment" data-fragment-index="3"><span class="math inline">\(\implies\)</span> The Negative Binomial Distribution</span></li>
<li><span class="math inline">\(\sigma^2\)</span>, <span class="math inline">\(\mu\)</span> will still be dependent!</li>
</ul></li>
</ul>
</div>
</section>
<section id="count-based-data-2" class="slide level2 slide-only unlisted">
<h2>Count-Based Data</h2>
<ul>
<li>We use the Negative Binomial distribution to model counts (<span class="math inline">\(y_{gi}\)</span>) for gene <span class="math inline">\(g\)</span> in sample <span class="math inline">\(i\)</span> <span class="citation" data-cites="Lun2016-vb">(<a href="#/section" role="doc-biblioref" onclick="">Lun, Chen, and Smyth 2016</a>)</span>
<ul>
<li>The expected counts <span class="math inline">\(E(y_{gi}) =\mu_{gi}\)</span></li>
<li>With overdispersed variance</li>
</ul></li>
</ul>
<p><span class="math display">\[
\text{var}(y_{gi}) = \sigma^2_g\left(\mu_{gi} +\phi\mu^2_{gi})\right), \text{where } \phi &gt; 0
\]</span></p>
<div class="fragment">
<ul>
<li>Can be thought of as like a Poisson Distribution with extra variation</li>
<li>Extra variation is strictly defined in quadratic relationship to mean
<ul>
<li>Sometimes described as a combination of <em>technical</em> &amp; <em>biological</em> variation</li>
</ul></li>
</ul>
</div>
</section>
<section id="count-based-data-3" class="slide level2 slide-only unlisted">
<h2>Count-Based Data</h2>
<ul>
<li>Fit NB <em>generalised linear models</em> (GLMs) to model counts and estimate logFC</li>
<li>Implemented in edgeR <span class="citation" data-cites="Chen2016-ig">(<a href="#/section" role="doc-biblioref" onclick="">Chen, Lun, and Smyth 2016</a>)</span> and DESeq2 <span class="citation" data-cites="love2014-ds">(<a href="#/section" role="doc-biblioref" onclick="">Love, Huber, and Anders 2014</a>)</span>
<ul>
<li>Slight differences in model-fitting</li>
<li>Overdispersion (<span class="math inline">\(\phi\)</span>) moderated in an analogous manner to variance for microarrays e.g.&nbsp;<span class="math inline">\(\phi = \phi(\mu_g)\)</span></li>
<li>Both default to FDR-adjusted <span class="math inline">\(p\)</span>-values</li>
</ul></li>
</ul>
</section>
<section id="count-based-data-4" class="slide level2 slide-only unlisted">
<h2>Count-Based Data</h2>
<ul>
<li>Poisson/NB-GLMs fit the rate of an event, i.e.&nbsp;counts per fixed measurement window</li>
<li>Sequencing data produces ‘libraries’ of counts <span class="math inline">\(\rightarrow\)</span> total counts = <em>library size</em></li>
<li>In model fitting <span class="math inline">\(\rightarrow\)</span> estimate rate as a function of library size</li>
</ul>
<div class="fragment">
<ul>
<li>TMM (edgeR) or RLE (DESeq2) approaches estimate scaling factors
<ul>
<li>Moderates the effect of highly expressed genes which dominate library size</li>
<li>e.g.&nbsp;Haemogloblin in blood samples can range between 30-60% of library</li>
<li>In a perfect world <span class="math inline">\(\implies\)</span> all scaling factors = 1</li>
</ul></li>
<li>Library size when modelling is multiplied by scaling factors
<ul>
<li>The most effective normalisation methods for RNA-Seq</li>
</ul></li>
</ul>
</div>
</section>
<section id="count-based-data-5" class="slide level2 slide-only unlisted">
<h2>Count-Based Data</h2>
<ul>
<li>A common alternative measure is <em>counts per million</em> (CPM) or logCPM
<ul>
<li>Mainly used for visualisation <em>not for analysis</em></li>
</ul></li>
<li>CPM is simply counts divided by (library size / 1,000,000)
<ul>
<li>No longer discrete <span class="math inline">\(\implies\)</span> continuous data</li>
</ul></li>
<li>Relationship between mean and variance <em>still retained</em> for logCPM
<ul>
<li>Can’t use naively in classic <code>limma</code>-based linear models</li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li>The limma-voom method <span class="citation" data-cites="Law2014-xq">(<a href="#/section" role="doc-biblioref" onclick="">Law et al. 2014</a>)</span> uses weights to break the mean-variance relationship
<ul>
<li>Can assume normally-distributed logCPM values</li>
</ul></li>
<li>Alternatively, the <code>limma-trend</code> works effectively</li>
</ul>
</div>
</section>
<section id="alternative-measure-for-rna-seq-counts" class="slide level2">
<h2>Alternative Measure for RNA-Seq Counts</h2>
<ul>
<li>Most designed to scale counts for length
<ul>
<li>Gene length is constant across samples <span class="math inline">\(\implies\)</span> not needed for analysis</li>
</ul></li>
<li>RPKM: Reads Per Kilobase of transcript per Million mapped reads
<ul>
<li>Dominated early RNA-Seq analyses <span class="math inline">\(\implies\)</span> <span class="text-red"><u>No longer in common use</u></span></li>
<li>Doesn’t handle different library compositions like TMM/RLE</li>
<li>Effectively makes RPKM not comparable across samples</li>
<li>FPKM was ‘Fragments Per Kilobase …’ <span class="math inline">\(\implies\)</span> paired reads</li>
</ul></li>
<li>TPM: Transcripts Per Million
<ul>
<li>Divides all counts by gene length</li>
<li>Scales across genes so library size is 10<sup>6</sup> reads</li>
<li>Only used for visualisation</li>
</ul></li>
</ul>
</section>
<section id="common-visualisation-techniques" class="slide level2">
<h2>Common Visualisation Techniques</h2>
<div class="columns">
<div class="column" style="width:55%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="assets/ma-plot-1.png"></p>
<figcaption>Taken from <span class="citation" data-cites="Sung2023-ob">Sung et al. (<a href="#/section" role="doc-biblioref" onclick="">2023</a>)</span></figcaption>
</figure>
</div>
</div><div class="column" style="width:45%;">
<ul>
<li>MA Plots show average expression (logCPM) against logFC</li>
<li>Sometimes called <em>smear</em> plots</li>
<li>Smoothed curve will highlight any bias</li>
<li>Low-signal genes usually very noisy</li>
</ul>
</div></div>
</section>
<section id="common-visualisation-techniques-1" class="slide level2 slide-only unlisted">
<h2>Common Visualisation Techniques</h2>
<div class="columns">
<div class="column" style="width:55%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="assets/volcano-plot-1.png"></p>
<figcaption>Taken from <span class="citation" data-cites="Sung2023-ob">Sung et al. (<a href="#/section" role="doc-biblioref" onclick="">2023</a>)</span></figcaption>
</figure>
</div>
</div><div class="column" style="width:45%;">
<ul>
<li>Volcano Plots show significance against logFC</li>
<li>Developed during the microarray era</li>
</ul>
</div></div>
</section></section>
<section>
<section id="alignment-to-a-reference-transcriptome" class="title-slide slide level1 center">
<h1>Alignment To A Reference Transcriptome</h1>

</section>
<section id="reference-transcriptomes" class="slide level2">
<h2>Reference Transcriptomes</h2>
<ul>
<li>An alternative to using a reference genome + gene annotations<br><em>align to a reference transcriptome</em></li>
<li>No longer need splice-aware aligners</li>
<li>Genome-based alignments (e.g.&nbsp;BAM files) no longer produced</li>
</ul>
<div class="fragment">
<ul>
<li>Many genes have multiple isoforms which share stretches of the same sequence</li>
<li>Reads will commonly align to multiple transcripts</li>
<li>Naive counting no longer viable</li>
</ul>
</div>
</section>
<section id="reference-transcriptomes-1" class="slide level2 slide-only unlisted">
<h2>Reference Transcriptomes</h2>
<ul>
<li>Two very similar approaches:
<ul>
<li><code>kallisto</code> <span class="citation" data-cites="Bray2016-qq">(<a href="#/section" role="doc-biblioref" onclick="">Bray et al. 2016</a>)</span> pseudo-aligns to a <em>de Bruijn</em> graph <span class="math inline">\(\implies\)</span> EM algorithm for counts</li>
<li><code>salmon</code> <span class="citation" data-cites="Patro2017-sn">(<a href="#/section" role="doc-biblioref" onclick="">Patro et al. 2017</a>)</span> aligns to reference <span class="math inline">\(\implies\)</span> EM/VBEM algorithm for counts</li>
</ul></li>
<li>Counts are fitted values (i.e.&nbsp;estimates) based on observed alignments
<ul>
<li>Are <em>transcript-level counts</em></li>
</ul></li>
<li>Alignments are downsampled &amp; bootstrapped <span class="math inline">\(\implies\)</span> estimate of confidence in counts</li>
</ul>
</section>
<section id="reference-transcriptomes-2" class="slide level2 slide-only unlisted">
<h2>Reference Transcriptomes</h2>
<ul>
<li>Many transcripts within a gene share multiple exons</li>
<li>Bootstraps provide confidence estimates for difficult/complicated transcripts</li>
</ul>

<img data-src="assets/gkv608fig1.jpeg" class="r-stretch quarto-figure-left"><p class="caption">Image taken from <span class="citation" data-cites="Yu2015-f5">Yu et al. (<a href="#/section" role="doc-biblioref" onclick="">2015</a>)</span></p></section>
<section id="differential-transcript-analysis" class="slide level2">
<h2>Differential Transcript Analysis</h2>
<ul>
<li>Dividing counts by bootstrap estimates <span class="math inline">\(\implies\)</span> Negative Binomial distribution <span class="citation" data-cites="Baldoni2024-zf">(<a href="#/section" role="doc-biblioref" onclick="">Baldoni et al. 2024</a>)</span>
<ul>
<li>Standard NB-GLMs applicable</li>
<li>Distribution not yet defined otherwise</li>
</ul></li>
<li>Transcript proportions within genes is alternative approach <span class="citation" data-cites="Soneson2015-hh">(<a href="#/section" role="doc-biblioref" onclick="">Soneson, Love, and Robinson 2015</a>)</span></li>
</ul>
<div class="fragment">
<ul>
<li>Summing transcript counts to gene counts <span class="math inline">\(\implies\)</span> no need for bootstrap estimates</li>
</ul>
</div>
</section></section>
<section>
<section id="references" class="title-slide slide level1 center">
<h1>References</h1>

</section>
<section id="section" class="slide level2 smaller scrollable">
<h2></h2>
<!-- This is where the bibliography gets injected -->
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-htseq2014" class="csl-entry" role="listitem">
Anders, Simon, Paul Theodor Pyl, and Wolfgang Huber. 2014. <span>“HTSeq—a Python Framework to Work with High-Throughput Sequencing Data.”</span> <em>Bioinformatics</em> 31 (2): 166–69. <a href="https://doi.org/10.1093/bioinformatics/btu638">https://doi.org/10.1093/bioinformatics/btu638</a>.
</div>
<div id="ref-Baldoni2024-zf" class="csl-entry" role="listitem">
Baldoni, Pedro L, Yunshun Chen, Soroor Hediyeh-Zadeh, Yang Liao, Xueyi Dong, Matthew E Ritchie, Wei Shi, and Gordon K Smyth. 2024. <span>“Dividing Out Quantification Uncertainty Allows Efficient Assessment of Differential Transcript Expression with <span class="nocase">edgeR</span>.”</span> <em>Nucleic Acids Res.</em> 52 (3): e13.
</div>
<div id="ref-Bray2016-qq" class="csl-entry" role="listitem">
Bray, Nicolas L, Harold Pimentel, Páll Melsted, and Lior Pachter. 2016. <span>“Near-Optimal Probabilistic <span>RNA</span>-Seq Quantification.”</span> <em>Nat. Biotechnol.</em> 34 (5): 525–27.
</div>
<div id="ref-Chen2016-ig" class="csl-entry" role="listitem">
Chen, Yunshun, Aaron T L Lun, and Gordon K Smyth. 2016. <span>“From Reads to Genes to Pathways: Differential Expression Analysis of <span>RNA</span>-Seq Experiments Using Rsubread and the <span class="nocase">edgeR</span> Quasi-Likelihood Pipeline.”</span> <em>F1000Res.</em> 5 (June): 1438.
</div>
<div id="ref-Dobin2013-us" class="csl-entry" role="listitem">
Dobin, Alexander, Carrie A Davis, Felix Schlesinger, Jorg Drenkow, Chris Zaleski, Sonali Jha, Philippe Batut, Mark Chaisson, and Thomas R Gingeras. 2013. <span>“<span>STAR</span>: Ultrafast Universal <span>RNA</span>-Seq Aligner.”</span> <em>Bioinformatics</em> 29 (1): 15–21.
</div>
<div id="ref-Kim2019-ib" class="csl-entry" role="listitem">
Kim, Daehwan, Joseph M Paggi, Chanhee Park, Christopher Bennett, and Steven L Salzberg. 2019. <span>“Graph-Based Genome Alignment and Genotyping with <span>HISAT2</span> and <span>HISAT</span>-Genotype.”</span> <em>Nat. Biotechnol.</em> 37 (8): 907–15.
</div>
<div id="ref-Law2014-xq" class="csl-entry" role="listitem">
Law, Charity W, Yunshun Chen, Wei Shi, and Gordon K Smyth. 2014. <span>“Voom: Precision Weights Unlock Linear Model Analysis Tools for <span>RNA</span>-Seq Read Counts.”</span> <em>Genome Biol.</em> 15 (2): R29.
</div>
<div id="ref-Li2011-iu" class="csl-entry" role="listitem">
Li, Bo, and Colin N Dewey. 2011. <span>“<span>RSEM</span>: Accurate Transcript Quantification from <span>RNA</span>-Seq Data with or Without a Reference Genome.”</span> <em>BMC Bioinformatics</em> 12 (August): 323.
</div>
<div id="ref-Liao2014-gy" class="csl-entry" role="listitem">
Liao, Yang, Gordon K Smyth, and Wei Shi. 2014. <span>“<span class="nocase">featureCounts</span>: An Efficient General Purpose Program for Assigning Sequence Reads to Genomic Features.”</span> <em>Bioinformatics</em> 30 (7): 923–30.
</div>
<div id="ref-love2014-ds" class="csl-entry" role="listitem">
Love, Michael I., Wolfgang Huber, and Simon Anders. 2014. <span>“Moderated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.”</span> <em>Genome Biology</em> 15: 550. <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
<div id="ref-Lun2016-vb" class="csl-entry" role="listitem">
Lun, Aaron T L, Yunshun Chen, and Gordon K Smyth. 2016. <span>“It’s <span>DE</span>-Licious: A Recipe for Differential Expression Analyses of <span>RNA</span>-Seq Experiments Using Quasi-Likelihood Methods in <span class="nocase">edgeR</span>.”</span> <em>Methods Mol. Biol.</em> 1418: 391–416.
</div>
<div id="ref-Patro2017-sn" class="csl-entry" role="listitem">
Patro, Rob, Geet Duggal, Michael I Love, Rafael A Irizarry, and Carl Kingsford. 2017. <span>“Salmon Provides Fast and Bias-Aware Quantification of Transcript Expression.”</span> <em>Nat. Methods</em> 14 (4): 417–19.
</div>
<div id="ref-Soneson2015-hh" class="csl-entry" role="listitem">
Soneson, Charlotte, Michael I Love, and Mark D Robinson. 2015. <span>“Differential Analyses for <span>RNA</span>-Seq: Transcript-Level Estimates Improve Gene-Level Inferences.”</span> <em>F1000Res.</em> 4 (1521): 1521.
</div>
<div id="ref-Sung2023-ob" class="csl-entry" role="listitem">
Sung, Jonghoo, Kate R Barratt, Stephen M Pederson, Chantal Chenu, Ines Reichert, Gerald J Atkins, Paul H Anderson, and Peter J Smitham. 2023. <span>“Unbiased Gene Expression Analysis of the Delayed Fracture Healing Observed in Zucker Diabetic Fatty Rats.”</span> <em>Bone Joint Res.</em> 12 (10): 657–66.
</div>
<div id="ref-Wang2009-hf" class="csl-entry" role="listitem">
Wang, Zhong, Mark Gerstein, and Michael Snyder. 2009. <span>“<span>RNA</span>-Seq: A Revolutionary Tool for Transcriptomics.”</span> <em>Nat. Rev. Genet.</em> 10 (1): 57–63.
</div>
<div id="ref-Yu2015-f5" class="csl-entry" role="listitem">
Yu, Nancy Yiu-Lin, Björn M. Hallström, Linn Fagerberg, Fredrik Ponten, Hideya Kawaji, Piero Carninci, Alistair R. R. Forrest, et al. 2015. <span>“Complementing Tissue Characterization by Integrating Transcriptome Profiling from the Human Protein Atlas and from the FANTOM5 Consortium.”</span> <em>Nucleic Acids Research</em> 43 (14): 6787–98. <a href="https://doi.org/10.1093/nar/gkv608">https://doi.org/10.1093/nar/gkv608</a>.
</div>
</div>

</section></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="rnaseq_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="rnaseq_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="rnaseq_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="rnaseq_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="rnaseq_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="rnaseq_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="rnaseq_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="rnaseq_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="rnaseq_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="rnaseq_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    <div id="logo-tki" style="position: fixed; z-index: 1000;">
      <img src="https://www.thekids.org.au/Static/Theme/images/TheKids-RGB_Logo_Colour-Pattern.svg" style="height: 80px;">
    </div>
    <div id="logo-bodl" style="position: fixed; z-index: 1000;">
      <img src="assets/bodl_logo_white_background.jpg" style="height: 80px;">
    </div>
    <div id="logo-uoa" style="z-index: 1000;">
      <img src="assets/UoA_logo_col_vert.png" style="height: 80px;">
    </div>


    <script>
    function positionLogos() {
      const tkiLogo = document.getElementById('logo-tki');
      const uoaLogo = document.getElementById('logo-uoa');
      const bodlLogo = document.getElementById('logo-bodl');

      // Check if this is RevealJS
      const isRevealJS = document.body.classList.contains('reveal') ||
                         document.querySelector('.reveal') !== null ||
                         typeof Reveal !== 'undefined';

      if (isRevealJS) {
        // RevealJS - both logos at bottom
        tkiLogo.style.bottom = '10px';
        tkiLogo.style.right = '340px';
        tkiLogo.style.top = 'auto';
        tkiLogo.style.display = 'block';

        bodlLogo.style.bottom = '10px';
        bodlLogo.style.right = '160px';
        bodlLogo.style.top = 'auto';
        bodlLogo.style.display = 'block';

        uoaLogo.style.bottom = '10px';
        uoaLogo.style.right = '40px';
        uoaLogo.style.top = 'auto';
        uoaLogo.style.display = 'block';
        uoaLogo.style.position = 'fixed';
      } else {
        // Flat HTML - only right logo at top
        tkiLogo.style.display = 'none';
        bodlLogo.style.display = 'none';

        uoaLogo.style.top = '30px';
        uoaLogo.style.left = '60%';
        uoaLogo.style.bottom = 'auto';
        uoaLogo.style.position = 'absolute';
        uoaLogo.style.display = 'block';
      }
    }

    // Run positioning
    document.addEventListener('DOMContentLoaded', positionLogos);
    setTimeout(positionLogos, 100);
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
            const codeEl = trigger.previousElementSibling.cloneNode(true);
            for (const childEl of codeEl.children) {
              if (isCodeAnnotation(childEl)) {
                childEl.remove();
              }
            }
            return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp('/' + window.location.host + '/');
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    

</body></html>